prymer/api/primer_pair.py

```diff
@@ -84,7 +84,11 @@ class PrimerPair(OligoLike):
     def __post_init__(self) -> None:
         # Derive the amplicon from the left and right primers. This must be done before
         # calling super() as `PrimerLike.id` depends on the amplicon being set
-        object.__setattr__(self, "_amplicon", self._calculate_amplicon())
+        object.__setattr__(
+            self,
+            "_amplicon",
+            PrimerPair.calculate_amplicon_span(self.left_primer, self.right_primer),
+        )
```

Comment:
**suggestion** Is this a good opportunity to refactor to make `amplicon` a `cached_property`, rather than mucking about with a private attribute and `setattr`?

e.g.

```py
@cached_property
def amplicon(self) -> Span:
    """The interval spanned by the pair's amplicon."""
    return self.calculate_amplicon_span(self.left_primer, self.right_primer)

Reply:
I'd really like to remove `_amplicon` as a "private" field.

As a consequence of the current implementation, both `asdict()` and `fields()` return a field that isn't accepted by the class constructor. This is unusual behavior and requires the user to remember and manually remove the field.

Could we make `amplicon` a `cached_property`, and update the post-init checks to reference that property? That protects all the cases I'm aware of motivating the current implementation:
1. Mutation-free access to an `amplicon` "attribute".
2. The value of `amplicon` is still derived from the input primers.

As a bonus, it removes the need to use `setattr` (currently necessary because we're trying to mutate a frozen dataclass after instantiation).

Reply:
I tried ... but failed.  Changed it to a cached property and tried calling it from post_init() to run the validation.  Now I get this in tests, which is super unhelpful and I'm not sure why:

```
ERROR tests/api/test_primer_pair.py - TypeError: No '__dict__' attribute on 'PrimerPair' instance to cache 'amplicon' property.
ERROR tests/api/test_primer_pair.py - TypeError: No '__dict__' attribute on 'PrimerPair' instance to cache 'amplicon' property.```

Going to merge without making the change, and we can circle back to it separately.